source/mesh_refinement/boundary.cc

```diff
@@ -0,0 +1,126 @@
+                             Patterns::List (Patterns::Anything()),
+                             "A comma separated list of names denoting those boundaries "
+                             "where there should be mesh refinement."
+                             "\n\n"
+                             "The names of the boundaries listed here can either be "
+                             "numbers (in which case they correspond to the numerical "
+                             "boundary indicators assigned by the geometry object), or they "
+                             "can correspond to any of the symbolic names the geometry object "
+                             "may have provided for each part of the boundary. You may want "
+                             "to compare this with the documentation of the geometry model you "
+                             "use in your model.");
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+    template <int dim>
+    void
+    Boundary<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Mesh refinement");
+      {
+        prm.enter_subsection("Boundary");
+        {
+          const std::vector<types::boundary_id> x_boundary_refinement_indicators
+            = this->get_geometry_model().translate_symbolic_boundary_names_to_ids(Utilities::split_string_list
+                                                                                  (prm.get ("Boundary refinement indicators")));
+          boundary_refinement_indicators
+            = std::set<types::boundary_id> (x_boundary_refinement_indicators.begin(),
+                                            x_boundary_refinement_indicators.end());
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+    template <int dim>
+    void
+    Boundary<dim>::execute(Vector<float> &indicators) const
+    {
+      indicators = 0;
+
+      // iterate over all of the cells and choose the ones at the indicated
+      // boundaries for refinement (assign the largest error to them)
+
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+
+      unsigned int i=0;
+      for (; cell!=endc; ++cell, ++i)
+        if (cell->is_locally_owned() && cell->at_boundary())
+          for (unsigned int face_no=0; face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
+            if (cell->face(face_no)->at_boundary())
+              {
+                const types::boundary_id boundary_indicator
+#if DEAL_II_VERSION_GTE(8,3,0)
+                  = cell->face(face_no)->boundary_id();
+#else
+                  = cell->face(face_no)->boundary_indicator();
+#endif
+                if ( boundary_refinement_indicators.find(boundary_indicator) !=
+                     boundary_refinement_indicators.end() )
+                  indicators(i) = 1.0;
```

Comment:
if you found a cell of interest, you may as well `break` out of the loop over the faces.